### PR TITLE
Initial Config Sync PR

### PR DESCRIPTION
**Pull request by mitchellh**
This is a **big PR** but only contains the initial implementation of config syncing:

* `internal/ceb` changes to sync config
* `internal/config` to parse `config` blocks
* `internal/server` changes to process and update configuration
* MVP of `vault` config sourcer but this requires Vault env vars for auth currently
* `kubernetes` config sourcer which works pretty well

## What this PR does NOT have

This PR is missing a lot of features we still want to add:

* CLI for just updating the config settings
* All the machinery to configure auth and such for sourcers
* Plugins for config sourcers rather than hard-coding them

### CODE REVIEW

File: builtin/k8s/config_sourcer.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	// refreshPeriod is the interval between refreshing secret values that
+	// aren't renewable or have a lease associated with them. If a Read is
+	// called again during this period, we will return cached values.
+	//
+	// We have to do this because Vault doesn't support any sort of blocking
+	// queries so unless it tells us a lease period, we can't safely refresh.
+	refreshPeriod = 30 * time.Second
+)
+
+// ConfigSourcer implements component.ConfigSourcer for Vault
```
Comment (krantzinator) [inline builtin/k8s/config_sourcer.go:32]:
```suggestion
// ConfigSourcer implements component.ConfigSourcer for K8s
```

**Review by krantzinator — APPROVED** [review]
The walkthrough + test coverage really helped fill out the story here, so thanks for all of that! There's a lot of cool stuff happening :tada:

File: internal/ceb/app_config.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+var (
+	// appConfigRefreshPeriod is the interval between checking for new
+	// config values. In a steady state, configuration NORMALLY doesn't
+	// change so this is set fairly high to avoid unnecessary load on
+	// dynamic config sources.
+	//
+	// NOTE(mitchellh): In the future, we'd like to build a way for
```
[the remark this answers is not in the document]
Comment (mitchellh) [reply internal/ceb/app_config.go:28]: I think not worth creating an issue, we should discuss internally if we want a backlog somewhere (I don't think GH issues are necessarily right).

**Review by briancain — APPROVED** [review]

**Review by evanphx — APPROVED** [review]
Probably should figure out freeport being not copied into here, but that's all!

**Pull request merged**